### PR TITLE
Major data structure changes

### PR DESCRIPTION
**Pull request by zjaxxon**
Please check for the major changes:
1. Major changes in data structure of CvoPointCloud
2. Major changes to fit the data structure in all constructors of CvoPointCloud
3. Labels function are disabled because of complex/unclear codes in contructors, marked TODO
4. Minor changes to introduce new main file and script for running aligements on stereo Tartanair 

Before Merging:
1. Cleaned old data structure
2. Minor changes in main files

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -76,7 +76,7 @@ add_definitions(-DOPENMP)
 #)
 
 find_package(CUDA)
-set (CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS}  --expt-extended-lambda  -gencode arch=compute_61,code=sm_61  -gencode arch=compute_52,code=sm_52 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_86,code=sm_86 "  )
+#set (CMAKE_CUDA_FLAGS "${CMAKE_CUDA_FLAGS}  --expt-extended-lambda  -gencode arch=compute_61,code=sm_61  -gencode arch=compute_52,code=sm_52 -gencode arch=compute_75,code=sm_75 -gencode arch=compute_86,code=sm_86 "  )
```
Comment (zeroAska) [inline CMakeLists.txt]: Have you tried compiling with these cuda arch flags?

File: include/UnifiedCvo/utils/CvoPointCloud.hpp [hunk 2]
```diff
@@ -170,14 +181,16 @@ namespace cvo {
...
+
+    std::vector<cvo::CvoPoint> points_;
     //std::vector<Eigen::Vector3f, Eigen::aligned_allocator<Eigen::Vector3f>> positions_;  // points position. x,y,z
     std::vector<Eigen::Vector3f> positions_; 
     //Eigen::Matrix<float, Eigen::Dynamic, 3> positions_;
     Eigen::MatrixXf features_;
     //Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic> features_;   // rgb, gradient in [0,1]
     //std::vector<Eigen::Matrix<float, 1, Eigen::Dynamic>> features_;   // rgb, gradient in [0,1]
     //Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic> normals_;  // surface normals
+    //TODO: fix labels here
```
Comment (zeroAska) [inline include/UnifiedCvo/utils/CvoPointCloud.hpp]: The posiitons_, features_, and labels_ can be removed in the future

File: include/UnifiedCvo/utils/CvoPointCloud.hpp [hunk 3]
```diff
@@ -129,17 +131,26 @@ namespace cvo {
...
+    std::vector<cvo::CvoPoint> get_points() const {return points_;}
+    cvo::CvoPoint point_at(unsigned int index) const {return points_[index];}
     int num_points() const {return num_points_;}
     int size() const {return num_points_;}
     int num_classes() const {return num_classes_;}
     int num_features() const {return feature_dimensions_;}
     int feature_dimensions() const {return feature_dimensions_;}
     int num_geometric_types() const {return num_geometric_types_; }
-    const std::vector<Eigen::Vector3f> & positions() const {return positions_;}
+    const std::vector<Eigen::Vector3f> positions() const {
```
Comment (zeroAska) [inline include/UnifiedCvo/utils/CvoPointCloud.hpp:142]: is this positions() still used?

File: CMakeLists.txt (hunk 1 above)
Comment (zjaxxon) [reply CMakeLists.txt]: Yes, this was causing a problem as I found the record. Might becuase of a merging issue. Successful on uncomment line 79

File: include/UnifiedCvo/utils/CvoPointCloud.hpp (hunk 3 above)
Comment (zjaxxon) [reply include/UnifiedCvo/utils/CvoPointCloud.hpp:142]: Yes, still in this version. Though it's better to replace this getter in all implementations I think. I will try on spare time.

File: include/UnifiedCvo/utils/CvoPointCloud.hpp (hunk 2 above)
Comment (zjaxxon) [reply include/UnifiedCvo/utils/CvoPointCloud.hpp]: I have checked that there were no calls to these variables in cvocloudpoint.cpp and cvopoint.cpp in this version. I will remove them in the next submit

File: src/utils/CvoPointCloud.cpp [hunk 4]
```diff
@@ -438,6 +438,9 @@ namespace cvo{
     //positions_.resize(output_uv.size());
     Mat33f intrinsic = calib.intrinsic();
 
+    num_classes_ = raw_image.num_classes();
+    num_geometric_types_ = 2;
+    feature_dimensions_ = raw_image.channels() + 2;
```
Comment (zeroAska) [inline src/utils/CvoPointCloud.cpp:478]: Just to make sure, if the new CvoPointCloud a template class with template argument num_class and feature_dim? If so we don't need to assign it to be  raw_image.channels() + 2;

File: src/utils/CvoPointCloud.cpp [hunk 5]
```diff
@@ -720,6 +724,9 @@ namespace cvo{
     //bool is_recording_depth_map = true;
     //static unsigned int depth_map_counter = 0;
     //positions_.resize(pre_depth_selected_ind.size());
+    num_classes_ = raw_image.num_classes();
+    num_geometric_types_ = 2;
+    feature_dimensions_ = raw_image.channels() + 2;
```
Comment (zeroAska) [inline src/utils/CvoPointCloud.cpp:740]: (Same comment above) Just to make sure, if the new CvoPointCloud a template class with template argument num_class and feature_dim? If so we don't need to assign it to be raw_image.channels() + 2;

File: src/utils/CvoPoint.cpp [hunk 6]
```diff
@@ -29,10 +29,10 @@ namespace cvo {
 //        features_(i, j) = pc[i].features[j];
           point.features[j] = pc[i].features[j];
       }
-      //TODO:fix label
-//      for (int j = 0 ; j < NUM_CLASSES; j++) {
-//        point.label_distribution[j] = pc[i].label_distribution[j];
-//      }
+
```
Comment (zeroAska) [inline src/utils/CvoPoint.cpp:32]: To make sure, can we move all functions related to pcl into a new separate cpp file? The reason is that we will want to allow users to decide if they want to compile with pcl themselves.

File: include/UnifiedCvo/utils/CvoPointCloud.hpp [hunk 7]
```diff
@@ -147,15 +147,15 @@ namespace cvo {
       return positions;
     }
     Eigen::Vector3f at(unsigned int index) const;
-    const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic> & labels() const { return labels_;}
-    Eigen::VectorXf label_at(unsigned int index) const { return labels_.row(index); }
+//    const Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic> & labels() const { return labels_;}
+    Eigen::VectorXf label_at(unsigned int index) const & { return Eigen::Map<const Eigen::VectorXf>(points_[index].label_distribution, NUM_CLASSES); }
```
Comment (zeroAska) [inline include/UnifiedCvo/utils/CvoPointCloud.hpp]: Just to make sure, do you think it is 
 Eigen::VectorXf label_at(unsigned int index) const & { return Eigen::Map<const Eigen::VectorXf>(points_[index].label_distribution, NUM_CLASSES); }

or 

const Eigen::VectorXf &  label_at(unsigned int index) const { return Eigen::Map<const Eigen::VectorXf>(points_[index].label_distribution, NUM_CLASSES); }

Comment (zjaxxon) [reply include/UnifiedCvo/utils/CvoPointCloud.hpp]: Thanks for pointing out! I believe that's my mistake. Fixed.

**Review by zeroAska — APPROVED** [review]
Good

**Comment by zjaxxon** [comment]
I have merged several of @zeroAska changes in the recent month and resolved several conflicts in Cmakelist and CvoPointCloud.cpp. The branch has no conflicts now.

File: src/utils/CvoPointCloud.cpp
Comment (zeroAska) [inline src/utils/CvoPointCloud.cpp:1]: Good

**Comment by zeroAska** [comment]
The updates look great. One more request before merging: would you like to provide a registration result comparison between the dev branch and the branch-to-merge for 

1. Demo: `./build/bin/cvo_align_gpu_two_color_pcd  demo_data/source.pcd  demo_data/target.pcd  cvo_params/cvo_outdoor_params.yaml`. 
2. Kitti Stereo sequence 05: 
```
    ./build/bin/cvo_align_gpu_semantic_img [kitti-repo-dir-on-server]/05 cvo_params/cvo_semantic_params_img_oct13.yaml \
                                       cvo_traj_semantic_05.txt  0 20000

```

Feel free to test it on our server. We expect the resulting pose matrices to be identical. Please give the resulting pose files for both branches. If there is any randomness, feel free to use a fixed random generator

**Comment by zjaxxon** [comment]
Hi Ray, please help me take a look at the most recent commit, which modified CvoGPU_impl.cu line 233-236 to fix the copy problem. Thanks
8-9 update: solved

**Pull request merged**